src/serializer/ResidualHeapSerializer.js

```diff
@@ -1279,9 +1420,30 @@ export class ResidualHeapSerializer {
           proto !== this.realm.intrinsics.ObjectPrototype &&
           this._findLastObjectPrototype(val) === this.realm.intrinsics.ObjectPrototype &&
           proto instanceof ObjectValue;
+        let isClass = false;
+        let classPrototype;
+        // if the object has a prototype that was a class (by check the constructor on the prototype)
+        if (proto instanceof ObjectValue && proto !== this.realm.intrinsics.ObjectPrototype) {
+          // check if this object is actually based from a class (by checking the constructor)
+          let valConstructor = Get(this.realm, val, "constructor");
```

Comment:
Not every instance of a class has a constructor. E.g.

```js
class Foo { }
delete Foo.prototype.constructor;
```

This could also be a getter which could be problematic.

The prototype on the constructor also not be the same anymore. E.g.

```js
class Foo {}
result = new Foo();
Foo.prototype = {};
```

or

```js
class Foo { };
result = Object.create({ constructor: Foo });
```

The only reason this needs to be serialized as a class is if any function remains that has this as the HomeObject. Otherwise it can just be a normal object.

Another solution might be to just tag the `ObjectValue` of the prototype at construction time just like the constructor is.

Reply:
I'll address this and the other concerns you had in a follow up tomorrow along with some tests. Thanks :)